File: react_native_app/src/features/productos/data/datasources/productoRemoteDataSource.ts
import { ApiClient } from '../../../../core/network/apiClient';
import { ProductoJson, ProductoModel } from '../models/ProductoModel';

interface ProductosListJson {
  products: ProductoJson[];
  total: number;
  skip: number;
  limit: number;
}

/** Acceso a la API REST de productos (DummyJSON). Solo conoce HTTP/JSON. */
export interface ProductoRemoteDataSource {
  getProductos(): Promise<ProductoModel[]>;
  getProducto(id: number): Promise<ProductoModel>;
  searchProductos(query: string): Promise<ProductoModel[]>;
  createProducto(producto: ProductoModel): Promise<ProductoModel>;
  updateProducto(producto: ProductoModel): Promise<ProductoModel>;
  deleteProducto(id: number): Promise<void>;
}

export class ProductoRemoteDataSourceImpl implements ProductoRemoteDataSource {
  constructor(private readonly client: ApiClient) {}

  async getProductos(): Promise<ProductoModel[]> {
    const json = await this.client.get<ProductosListJson>('/products?limit=30');
    return json.products.map(ProductoModel.fromJson);
  }

  async getProducto(id: number): Promise<ProductoModel> {
    const json = await this.client.get<ProductoJson>(`/products/${id}`);
    return ProductoModel.fromJson(json);
  }

  async searchProductos(query: string): Promise<ProductoModel[]> {
    const json = await this.client.get<ProductosListJson>(
      `/products/search?q=${encodeURIComponent(query)}`,
    );
    return json.products.map(ProductoModel.fromJson);
  }

  async createProducto(producto: ProductoModel): Promise<ProductoModel> {
    const json = await this.client.post<ProductoJson>(
      '/products/add',
      producto.toJson(),
    );
    return ProductoModel.fromJson(json);
  }

  async updateProducto(producto: ProductoModel): Promise<ProductoModel> {
    const json = await this.client.put<ProductoJson>(
      `/products/${producto.id}`,
      producto.toJson(),
    );
    return ProductoModel.fromJson(json);
  }

  async deleteProducto(id: number): Promise<void> {
    await this.client.delete<ProductoJson>(`/products/${id}`);
  }
}

File: flutter_app/lib/features/productos/data/datasources/producto_remote_datasource.dart
import '../../../../core/network/api_client.dart';
import '../models/producto_model.dart';

/// Acceso a la API REST de productos (DummyJSON). Solo conoce HTTP/JSON.
abstract interface class ProductoRemoteDataSource {
  Future<List<ProductoModel>> getProductos();
  Future<ProductoModel> getProducto(int id);
  Future<List<ProductoModel>> searchProductos(String query);
  Future<ProductoModel> createProducto(ProductoModel producto);
  Future<ProductoModel> updateProducto(ProductoModel producto);
  Future<void> deleteProducto(int id);
}

class ProductoRemoteDataSourceImpl implements ProductoRemoteDataSource {
  final ApiClient client;
  ProductoRemoteDataSourceImpl(this.client);

  @override
  Future<List<ProductoModel>> getProductos() async {
    final json = await client.get('/products?limit=30');
    final list = (json['products'] as List).cast<Map<String, dynamic>>();
    return list.map(ProductoModel.fromJson).toList();
  }

  @override
  Future<ProductoModel> getProducto(int id) async {
    final json = await client.get('/products/$id');
    return ProductoModel.fromJson(json as Map<String, dynamic>);
  }

  @override
  Future<List<ProductoModel>> searchProductos(String query) async {
    final json = await client.get('/products/search?q=${Uri.encodeQueryComponent(query)}');
    final list = (json['products'] as List).cast<Map<String, dynamic>>();
    return list.map(ProductoModel.fromJson).toList();
  }

  @override
  Future<ProductoModel> createProducto(ProductoModel producto) async {
    final json = await client.post('/products/add', producto.toJson());
    return ProductoModel.fromJson(json as Map<String, dynamic>);
  }

  @override
  Future<ProductoModel> updateProducto(ProductoModel producto) async {
    final json = await client.put('/products/${producto.id}', producto.toJson());
    return ProductoModel.fromJson(json as Map<String, dynamic>);
  }

  @override
  Future<void> deleteProducto(int id) async {
    await client.delete('/products/$id');
  }
}
